### PR TITLE
Generative Model for OASIS Brain Dataset using VQVAE 

### PR DESCRIPTION
**Pull request by tara-bash**
This pull request introduces a generative model utilizing VQVAE for the OASIS brain dataset. The primary objective was to achieve a "reasonably clear image" and a Structured Similarity (SSIM) score exceeding 0.6.

**Changes made:** 

- Add Modules.py: This file contains essential helper functions and the main VQVAE model classes, including the encoder, decoder, and the vector quantization process.
- Add Dataset.py: Manages data preprocessing for the OASIS brain dataset and initializing data loaders for training, validation, and testing. It uses a batch size of 32 and grayscale normalization.
- Add Train.py: This script manages the core training loop for the VQVAE on the OASIS dataset. It handles training, validation, and testing of the VQ-VAE model and saves model checkpoints based on improvements in the combined metric of SSIM and reconstruction loss.
- Add Predict.py: Demonstrates the usage of the trained model. It reads saved model parameters from the best epoch, loads the corresponding pre-trained VQVAE model, and generates and saves image samples from the test data.

**Model Components:**

-	VQEmbedding: A module to manage the vector quantization of embeddings.
-	ResBlock: A residual block component designed for efficiency in the neural network.
-	VectorQuantizedVAE: The main VQ-VAE class comprising the encoder, the decoder, and the integration of the above components. It facilitates the end-to-end process from encoding an image to producing its reconstructed version.

Ensure that PyTorch, torchvision, numpy, matplotlib, and skimage libraries are installed. Whilst not a necessary, it is recommended that a GPU is available for training the model as utilising the GPU will significantly expedite the training process of the model. The OASIS brain dataset is not included in the repository and will need to be downloaded externally to the project. After obtaining the dataset, adjust the paths within the dataset.py file to correctly point to the training, validation, and test datasets. 

Thank you for taking the time to review my repository. 

Tara Bashirzadeh (s47425055)

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme:
- Project Overview: Good
- Model: VQ OK, but no details on prior
- Data + Preprocessing: Good
- Usage: Need code usage ie which command(s) to run
- Training / Loss Curve(s): Good
- Result Demonstration: Good
- References: Good

Functionality:
- Reconstruction: Good
- **Generation: Not completed**

Code:
- Consistent with Results: Yes
- File Structure: Good
- Commenting: Good
- Commit frequency: Good
- Commit messages: Good
- Note: No code for prior found

Other Comments:
N/A

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem no generation -1
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage (missing generation results) -1
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable/good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
